File: apps/ios/Sources/BikeCore/WorkspaceJSON.swift
import Foundation

public struct WorkspacePayload: Equatable, Sendable {
    public var workspace: Workspace
    public var raw: [String: JSONValue]
    public var recovery: WorkspaceRecovery?

    public init(workspace: Workspace, raw: [String: JSONValue], recovery: WorkspaceRecovery? = nil) {
        self.workspace = workspace
        self.raw = raw
        self.recovery = recovery
    }
}

public struct WorkspaceRecovery: Equatable, Sendable {
    public var backupFileName: String

    public init(backupFileName: String) {
        self.backupFileName = backupFileName
    }
}

public enum WorkspaceJSON {
    public static let decoder: JSONDecoder = JSONDecoder()

    public static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return encoder
    }()

    public static func decode(_ source: String) throws -> WorkspacePayload {
        guard let data = source.data(using: .utf8) else {
            throw WorkspaceError.invalidWorkspace
        }
        let rawValue = try decoder.decode(JSONValue.self, from: data)
        guard let raw = rawValue.objectValue else {
            throw WorkspaceError.invalidWorkspace
        }
        let workspace = try normalize(decoder.decode(Workspace.self, from: data))
        return WorkspacePayload(workspace: workspace, raw: raw)
    }

    public static func encode(_ workspace: Workspace) throws -> String {
        let data = try encoder.encode(workspace)
        return String(decoding: data, as: UTF8.self)
    }

    public static func encode(_ payload: WorkspacePayload) throws -> String {
        let encodedData = try encoder.encode(payload.workspace)
        let encodedValue = try decoder.decode(JSONValue.self, from: encodedData)
        guard let encoded = encodedValue.objectValue else {
            throw WorkspaceError.invalidWorkspace
        }
        let merged = mergeWorkspace(raw: payload.raw, encoded: encoded)
        let data = try encoder.encode(JSONValue.object(merged))
        return String(decoding: data, as: UTF8.self)
    }

    public static func payload(for workspace: Workspace) throws -> WorkspacePayload {
        let data = try encoder.encode(workspace)
        let raw = try decoder.decode(JSONValue.self, from: data).objectValue ?? [:]
        return WorkspacePayload(workspace: workspace, raw: raw)
    }

    private static func normalize(_ workspace: Workspace) throws -> Workspace {
        guard !workspace.documents.isEmpty else {
            throw WorkspaceError.invalidWorkspace
        }
        if workspace.documents.contains(where: { $0.id == workspace.activeDocumentId }) {
            return workspace
        }
        var next = workspace
        next.activeDocumentId = workspace.documents[0].id
        return next
    }

    private static func mergeWorkspace(
        raw: [String: JSONValue],
        encoded: [String: JSONValue]
    ) -> [String: JSONValue] {
        var result = raw
        result["version"] = encoded["version"]
        result["activeDocumentId"] = encoded["activeDocumentId"]
        result["documents"] = mergeDocuments(
            raw: raw["documents"]?.arrayValue,
            encoded: encoded["documents"]?.arrayValue ?? []
        )
        return result
    }

    private static func mergeDocuments(raw: [JSONValue]?, encoded: [JSONValue]) -> JSONValue {
        let rawById = Dictionary(
            uniqueKeysWithValues: raw?
                .compactMap { value -> (String, [String: JSONValue])? in
                    guard let object = value.objectValue, let id = object["id"]?.stringValue else { return nil }
                    return (id, object)
                } ?? []
        )
        return .array(encoded.map { value in
            guard let encodedDocument = value.objectValue else { return value }
            let id = encodedDocument["id"]?.stringValue
            return .object(mergeDocument(raw: id.flatMap { rawById[$0] }, encoded: encodedDocument))
        })
    }

    private static func mergeDocument(
        raw: [String: JSONValue]?,
        encoded: [String: JSONValue]
    ) -> [String: JSONValue] {
        var result = raw ?? [:]
        putKnown(encoded: encoded, into: &result, keys: documentKeys)
        result["nodes"] = mergeNodes(
            raw: raw?["nodes"]?.arrayValue,
            encoded: encoded["nodes"]?.arrayValue ?? []
        )
        return result
    }

    private static func mergeNodes(raw: [JSONValue]?, encoded: [JSONValue]) -> JSONValue {
        let rawById = Dictionary(
            uniqueKeysWithValues: raw?
                .compactMap { value -> (String, [String: JSONValue])? in
                    guard let object = value.objectValue, let id = object["id"]?.stringValue else { return nil }
                    return (id, object)
                } ?? []
        )
        return .array(encoded.map { value in
            guard let encodedNode = value.objectValue else { return value }
            let id = encodedNode["id"]?.stringValue
            return .object(mergeNode(raw: id.flatMap { rawById[$0] }, encoded: encodedNode))
        })
    }

    private static func mergeNode(
        raw: [String: JSONValue]?,
        encoded: [String: JSONValue]
    ) -> [String: JSONValue] {
        var result = raw ?? [:]
        putKnown(encoded: encoded, into: &result, keys: nodeKeys)
        result["children"] = mergeNodes(
            raw: raw?["children"]?.arrayValue,
            encoded: encoded["children"]?.arrayValue ?? []
        )
        return result
    }

    private static func putKnown(
        encoded: [String: JSONValue],
        into result: inout [String: JSONValue],
        keys: Set<String>
    ) {
        for key in keys {
            if let value = encoded[key] {
                result[key] = value
            } else {
                result.removeValue(forKey: key)
            }
        }
    }

    private static let documentKeys: Set<String> = [
        "id",
        "title",
        "createdAt",
        "updatedAt",
        "markdownSource",
        "markdownUpdatedAt",
        "isShortcut"
    ]

    private static let nodeKeys: Set<String> = [
        "id",
        "text",
        "note",
        "checked",
        "collapsed",
        "color",
        "headingLevel",
        "bold",
        "italic",
        "underline",
        "strike",
        "highlight",
        "icon",
        "imageName",
        "imageAlt",
        "table",
        "codeBlock",
        "codeLanguage",
        "isTodo"
    ]
}

File: apps/android/app/src/main/java/com/bike/android/data/WorkspaceJson.kt
package com.bike.android.data

import kotlinx.serialization.encodeToString
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject

data class WorkspacePayload(
    val workspace: Workspace,
    val raw: JsonObject,
    val recovery: WorkspaceRecovery? = null,
)

object WorkspaceJson {
    @OptIn(ExperimentalSerializationApi::class)
    val json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
        encodeDefaults = true
        prettyPrint = true
    }

    fun decode(source: String): WorkspacePayload {
        val raw = json.parseToJsonElement(source).jsonObject
        val workspace = json.decodeFromJsonElement<Workspace>(raw).normalized()
        return WorkspacePayload(
            workspace = workspace,
            raw = raw,
        )
    }

    fun encode(workspace: Workspace): String =
        json.encodeToString(workspace)

    fun encode(payload: WorkspacePayload): String {
        val encoded = json.encodeToJsonElement(payload.workspace).jsonObject
        val merged = mergeWorkspace(payload.raw, encoded)
        return json.encodeToString(JsonObject.serializer(), merged)
    }

    private fun mergeWorkspace(raw: JsonObject, encoded: JsonObject): JsonObject =
        JsonObject(
            raw.toMutableMap().apply {
                this["version"] = encoded.getValue("version")
                this["activeDocumentId"] = encoded.getValue("activeDocumentId")
                this["documents"] = mergeDocuments(
                    raw["documents"]?.jsonArrayOrNull(),
                    encoded.getValue("documents").jsonArray,
                )
            },
        )

    private fun mergeDocuments(raw: JsonArray?, encoded: JsonArray): JsonArray {
        val rawById = raw.orEmpty()
            .mapNotNull { it.jsonObjectOrNull() }
            .associateBy { it.stringValue("id") }

        return JsonArray(
            encoded.map { document ->
                val encodedDocument = document.jsonObject
                mergeDocument(rawById[encodedDocument.stringValue("id")], encodedDocument)
            },
        )
    }

    private fun mergeDocument(raw: JsonObject?, encoded: JsonObject): JsonObject =
        JsonObject(
            (raw?.toMutableMap() ?: mutableMapOf()).apply {
                putKnown(encoded, DOCUMENT_KEYS)
                this["nodes"] = mergeNodes(
                    raw?.get("nodes")?.jsonArrayOrNull(),
                    encoded.getValue("nodes").jsonArray,
                )
            },
        )

    private fun mergeNodes(raw: JsonArray?, encoded: JsonArray): JsonArray {
        val rawById = raw.orEmpty()
            .mapNotNull { it.jsonObjectOrNull() }
            .associateBy { it.stringValue("id") }

        return JsonArray(
            encoded.map { node ->
                val encodedNode = node.jsonObject
                mergeNode(rawById[encodedNode.stringValue("id")], encodedNode)
            },
        )
    }

    private fun mergeNode(raw: JsonObject?, encoded: JsonObject): JsonObject =
        JsonObject(
            (raw?.toMutableMap() ?: mutableMapOf()).apply {
                putKnown(encoded, NODE_KEYS)
                this["children"] = mergeNodes(
                    raw?.get("children")?.jsonArrayOrNull(),
                    encoded.getValue("children").jsonArray,
                )
            },
        )

    private fun MutableMap<String, JsonElement>.putKnown(
        encoded: JsonObject,
        keys: Set<String>,
    ) {
        keys.forEach { key ->
            if (encoded.containsKey(key)) {
                this[key] = encoded.getValue(key)
            } else {
                remove(key)
            }
        }
    }

    private fun JsonElement.jsonObjectOrNull(): JsonObject? =
        this as? JsonObject

    private fun JsonElement.jsonArrayOrNull(): JsonArray? =
        this as? JsonArray

    private fun JsonObject.stringValue(key: String): String =
        (this[key] as? JsonPrimitive)?.content.orEmpty()

    private fun Workspace.normalized(): Workspace {
        require(documents.isNotEmpty()) { "工作区至少需要一篇文档" }
        return if (documents.any { it.id == activeDocumentId }) {
            this
        } else {
            copy(activeDocumentId = documents.first().id)
        }
    }

    private val DOCUMENT_KEYS = setOf(
        "id",
        "title",
        "createdAt",
        "updatedAt",
        "markdownSource",
        "markdownUpdatedAt",
        "isShortcut",
    )

    private val NODE_KEYS = setOf(
        "id",
        "text",
        "note",
        "checked",
        "collapsed",
        "color",
        "headingLevel",
        "bold",
        "italic",
        "underline",
        "strike",
        "highlight",
        "icon",
        "imageName",
        "imageAlt",
        "table",
        "codeBlock",
        "codeLanguage",
        "isTodo",
    )
}
